lib/libc/minimal/source/math/sqrt.c

```diff
@@ -5,19 +5,50 @@
  */
 
 #include <math.h>
+#include <stdint.h>
+#include <zephyr/sys/util.h>
 
-double sqrt(double value)
+#define MAX_D_ITTERATIONS	8  /* usually converges in 5 loops */
+				   /* this ensures we break out of the loop */
+#define MAX_D_ERROR_COUNT	5  /* when result almost converges, stop */
+#define EXP_MASK64	GENMASK64(62, 52)
+
+double sqrt(double square)
 {
-	double sqrt = value / 3;
 	int i;
+	int64_t	exponent;
+	double root;
+	double last;
+	int64_t *p_square = (int64_t *)&square;
+	int64_t *p_root = (int64_t *)&root;
+	int64_t *p_last = (int64_t *)&last;
 
-	if (value <= 0) {
-		return 0;
+	if (square == 0.0) {
+		return square;
+	}
+	if (square < 0.0) {
+		return (square - square) / (square - square);
 	}
 
-	for (i = 0; i < 6; i++) {
-		sqrt = (sqrt + value / sqrt) / 2;
+	/* we need a good starting guess so that this will converge quickly,
+	 * we can do this by dividing the exponent part of the float by 2
+	 * this assumes IEEE-754 format doubles
+	 */
+	exponent = ((*p_square & EXP_MASK64)>>52)-1023;
+	if (exponent == 0x7FF-1023) {
+		/* the number is a NAN or inf, return NaN or inf */
+		return square + square;
```

Comment:
(and the trick here with the addition converts a signaling nan into a non-signaling nan and raise the exception).

Reply:
just at you had suggested ;-)

@keith-packard thank you for all of your great suggestions.